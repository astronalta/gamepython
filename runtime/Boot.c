/*
 * Copyright (c) 2010 Matt Fichman
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, APEXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 */

#include "String.h"
#include <stdio.h>
#include <stdlib.h>

void boot_print_str(String* string) {
    // Write string to stdout.  This is function is here for convenience's 
    // sake.  Once a full-fledged IO framework has been written, this function
    // will really only be useful for simple output.

    fwrite(string->data, 1, string->length, stdout);
    fflush(stdout);
}

void boot_print_int(Int integer) {
    // Print an integer to stdout.  This function is here only to run initial
    // tests on the compiler, and isn't part of the public API.

#ifdef DARWIN
    fprintf(stdout, "%lld", integer);
#else
    fprintf(stdout, "%ld", integer);    
#endif
    fflush(stdout);
}

void boot_print_char(Char character) {
    // Print a character to stdout.  This function is not part of the public 
    // API for the Apollo library.

    fputc(character, stdout);
    fflush(stdout);
}

void boot_dummy(int a, int b, int c) {
    boot_print_int(a);
    boot_print_int(b);
    boot_print_int(c);
}
