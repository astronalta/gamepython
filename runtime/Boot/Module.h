/*
 * Copyright (c) 2010 Matt Fichman
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, APEXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 */

#ifndef JOGO_BOOT_H
#define JOGO_BOOT_H

#include "Primitives.h"

Ptr Boot_malloc(Int size);
Ptr Boot_calloc(Int size);
void Boot_mzero(Ptr val, Int size);
void Boot_memcpy(Ptr src, Ptr dst, Int size);
void Boot_free(Ptr memory);
void Boot_abort();

#endif
