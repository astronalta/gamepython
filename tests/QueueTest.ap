import Boot

test_enq_deq1() {
    queue = Queue[String](0)
    i = 0
    while i < 10 {
        queue.enq(i.str)
        ++i
    }

    i = 0
    while i < 10 {
        val = queue.deq()
        Io::println(val)
        if val != i.str {
            Io::println('fail')
        }
        ++i
    }
    Io::println('test_enq_deq1 OK')
}

test_enq_deq2() {
    queue = Queue[String](0)
    i = 0
    while i < 5000 {
        queue.enq(i.str)
        ++i
    }
    i = 0
    while i < 5000 {
        val = queue.deq()
        if val != i.str {
            Io::println('fail')
        }
        ++i
    }
    Io::println('test_enq_deq2 OK')
}

test_enq_deq3() {
    queue = Queue[String](0)
    i = 0
    while i < 400 {
        queue.enq(i.str)
        val = queue.deq()
        if val != i.str {
            Io::println('fail')
        }
        ++i
    }
    Io::println('test_enq_deq3 OK')
}

test_enq_deq4() {
    queue = Queue[String](0)
    i = 0
    while i < 200 {
        j = 0
        while j < 18 {
            queue.enq(j.str)
            ++j
        }
        j = 0
        while j < 18 {
            if j.str != queue.deq() {
                Io::println('fail')
            }
            ++j
        }
        ++i
    }
    Io::println('test_enq_deq4 OK')
}

test_count() {
    queue = Queue[String](0)
    i = 0
    while i < 120 {
        queue.enq(i.str)
        ++i
    }
    Io::println(queue.count.str)
    queue.deq()
    Io::println(queue.count.str)
    Io::println('test_count OK')
}

test_first_last() {
    queue = Queue[String](0)
    queue.enq('one')
    queue.enq('two')
    queue.enq('three')
    Io::println(queue.last)
    Io::println(queue.first)
    queue.deq()
    Io::println(queue.last)
    Io::println(queue.first)
    Io::println('test_first_last OK')
}

test_iter() {
    queue = Queue[String](0)
    queue.enq('one')
    queue.enq('two')
    queue.enq('three')

    for item in queue {
        Io::println(item)
    }
    queue.deq()
    for item in queue {
        Io::println(item)
    }
}


main() Int {
    test_enq_deq1()
    test_enq_deq2()
    test_enq_deq3()
    test_enq_deq4()
    test_count()
    test_first_last()
    test_iter()
    ret 0
}
