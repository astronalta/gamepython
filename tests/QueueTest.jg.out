0
1
2
3
4
5
6
7
8
9
test_enq_deq1 OK
test_enq_deq2 OK
test_enq_deq3 OK
test_enq_deq4 OK
120
119
test_count OK
three
one
three
two
test_first_last OK
one
two
three
two
three
